R/vpc-simulation.R

```diff
@@ -0,0 +1,533 @@
+#'    - Adds a new custom `$SIMULATION` record using user specified values (e.g.
+#'     `seed` and `n`). `TRUE=FINAL` is appended to ensure the final values are
+#'     used rather than the initial estimates.
+#'    - Adds a new `$TABLE` record tabling out simulated values (`sim_cols`) and
+#'    `.join_col` column(s)
+#'    - Adds a new `$MSFI` record (run with `NOMSFTEST`) pointing to the `MSF`
+#'    file of `.mod`
+#'
+#' @seealso nm_join_sim
+#' @examples
+#' \dontrun{
+#'
+#' # Create new `nmsim` model object
+#' .sim <- new_sim_model(
+#'   .mod,
+#'   n = 500, seed = 1234,
+#'   .join_col = c("NUM", "ID")
+#' )
+#'
+#' # Store run details in specification file
+#' #  - this is why these functions aren't exported
+#' make_sim_spec(
+#'   .sim,
+#'   sim_args = list(n = 500, seed = 1234),
+#'  .join_col = c("NUM", "ID")
+#' )
+#'
+#' # Visualize changes
+#' model_diff(.sim)
+#'
+#' # Submit
+#' submit_model(.sim)
+#' }
+#'
+#' @return S3 object of class `bbi_nmsim_model`.
+#' @keywords internal
+new_sim_model <- function(
+    .mod,
+    n = 200,
+    seed = 1234,
+    data = NULL,
+    sim_cols = c("DV", "PRED"),
+    .join_col = "NUM",
+    .sim_dir = get_output_dir(.mod),
+    .inherit_tags = TRUE,
+    .overwrite = FALSE
+){
+  checkmate::assert_numeric(n, lower = 1)
+  checkmate::assert_numeric(seed)
+  check_model_object(.mod, NM_MOD_CLASS)
+  check_model_for_sim(.mod, .join_col)
+
+  # Create new simulation model object
+  model_dir <- get_model_working_directory(.mod)
+  sim_dir_rel <- fs::path_rel(.sim_dir, model_dir)
+  output_dir <- file.path(sim_dir_rel, glue("{get_model_id(.mod)}-sim"))
+
+  .sim_mod <- copy_model_from(
+    .parent_mod = .mod,
+    .new_model = output_dir,
+    .add_tags = "SIMULATION",
+    .inherit_tags = .inherit_tags,
+    .update_model_file = TRUE,
+    .overwrite = .overwrite
+  )
+
+  .sim_mod[[YAML_MOD_TYPE]] <- "nmsim"
+  .sim_mod <- save_model_yaml(.sim_mod)
+
+  # Overwrite problem statement
+  prob <- glue("Simulation of model {get_model_id(.mod)}")
+  modify_prob_statement(.sim_mod, prob)
+
+
+  # Handle input data and adjust $DATA record
+  if(!is.null(data)){
+    checkmate::assert_data_frame(data)
+    input_cols <- get_input_columns(.mod) # Taken from control stream
+    if(!all(input_cols %in% names(data))){
+      missing_cols <- input_cols[!(input_cols %in% names(data))]
+      missing_txt <- paste(missing_cols, collapse = ", ")
+      rlang::abort(
+        c(
+          glue("The following required input columns were not found in the input data: {missing_txt}"),
+          "Check `nm_data(.mod)` to see expected columns."
+        )
+      )
+    }
+
+    # Remove any extra columns
+    data <- dplyr::select(data, all_of(input_cols))
+
+    # Save data to output dir of .mod
+    data_path_new <- file.path(.sim_dir, "sim-data.csv")
+    readr::write_csv(data, data_path_new)
+    verbose_msg(glue("Saving input data to `{data_path_new}`"))
```

Comment:
I don't think we need the message here; but I'd log somewhere in the output files.